src/dataflow/src/source/postgres.rs

```diff
@@ -157,12 +169,34 @@ impl PostgresSourceReader {
                         let a: Datum = row.get(n).into();
                         a
                     }));
-                    snapshot_tx.insert(mz_row).await?;
+                    try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
+                    try_fatal!(buffer.write(&try_fatal!(bincode::serialize(&mz_row))).await);
+                    try_fatal!(buffer.write(&[b'\n']).await);
+                }
+            }
+        }
+        try_recoverable!(client.simple_query("COMMIT;").await);
+        Ok(())
+    }
+
+    /// Reverts a failed snapshot by deleting any processed rows from the dataflow.
+    async fn revert_snapshot(
+        &self,
+        snapshot_tx: &mut SourceTransaction<'_>,
+        buffer: &mut BufReader<BufWriter<File>>,
+    ) -> Result<(), anyhow::Error> {
+        buffer.seek(SeekFrom::Start(0)).await?;
+        loop {
+            let mut bytes = Vec::new();
+            match buffer.read_until(b'\n', &mut bytes).await? {
```

Comment:
This will split the buffer every time the byte 0x0a (\n) appears but because this is a binary encoding there is no guarantee that this will not appear in the data.

I think what we need to do instead is call `bincode::deserialize_from` in a loop and give it our reader directly. This way it will advance the reader based on the underlying encoding. This also assumes we do `bincode::serialize_into` on the writer in the other side.

@philip-stoev I believe this is the source of the errors you're seeing